### PR TITLE
*: support raft learner in etcd - part 1

### PR DESCRIPTION
**Pull request by jingyih**
The first 7 commits of #10645. Rebased to master.

cc @xiang90

### CODE REVIEW

File: clientv3/cluster.go [hunk 1]
```diff
@@ -20,29 +20,34 @@ import (
...
+	MemberPromoteResponse pb.MemberPromoteResponse
 )
 
 type Cluster interface {
 	// MemberList lists the current cluster membership.
 	MemberList(ctx context.Context) (*MemberListResponse, error)
 
 	// MemberAdd adds a new member into the cluster.
-	MemberAdd(ctx context.Context, peerAddrs []string) (*MemberAddResponse, error)
+	MemberAdd(ctx context.Context, peerAddrs []string, isLearner bool) (*MemberAddResponse, error)
```
Comment (xiang90) [inline clientv3/cluster.go]: as I mentioned in another PR. Probably we should keep this method, and add MemberAddAsLearner.

File: etcdserver/api/membership/member.go [hunk 2]
```diff
@@ -51,10 +53,13 @@ type Member struct {
 
 // NewMember creates a Member without an ID and generates one based on the
 // cluster name, peer URLs, and time. This is used for bootstrapping/adding new member.
-func NewMember(name string, peerURLs types.URLs, clusterName string, now *time.Time) *Member {
+func NewMember(name string, peerURLs types.URLs, clusterName string, now *time.Time, isLearner bool) *Member {
```
Comment (xiang90) [inline etcdserver/api/membership/member.go]: make this func newMember(..., isLearner bool). Let NewMemeber func call in newMember. Add NewMemberAsLearner.

**Comment by xiang90** [comment]
lgtm after improving the interface thing.

**Comment by xiang90** [comment]
build failed with

```
clientv3/integration/cluster_test.go:80:25: not enough arguments in call to capi.Cluster.MemberAdd
	have (context.Context, []string)
	want (context.Context, []string, bool)
FAIL	go.etcd.io/etcd/v3/clientv3/integration [build failed]
```

**Comment by jingyih** [comment]
> build failed with
> 
> ```
> clientv3/integration/cluster_test.go:80:25: not enough arguments in call to capi.Cluster.MemberAdd
> 	have (context.Context, []string)
> 	want (context.Context, []string, bool)
> FAIL	go.etcd.io/etcd/v3/clientv3/integration [build failed]
> ```

Yeah. That is known. There are conflicts to master branch due to changes added to master branch after learner feature branch's branching off. They are fixed after a later rebasing. This particular one is fixed in 11e41685030846ba495a4263d8db352d0e9fc7d0.

**Comment by jingyih** [comment]
There are also conflicts in functional tests, which are fixed in https://github.com/etcd-io/etcd/pull/10645/commits/c8fbc4d50ef97bc2ba78f0e71f8b9b78de0494ad.

Since we are going to have a new function `MemberAddAsLearner`, these conflicts will be fixed.

**Comment by jingyih** [comment]
BTW, we had travis CI enabled in learner feature branch during the development. So all tests will pass if all commits are included.

File: etcdserver/api/membership/member.go (hunk 2 above)
Comment (jingyih) [reply etcdserver/api/membership/member.go]: Done.

File: proxy/grpcproxy/cluster.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ import (
 	"go.etcd.io/etcd/v3/etcdserver/api/v3rpc/rpctypes"
 	pb "go.etcd.io/etcd/v3/etcdserver/etcdserverpb"
 
+	"errors"
```
Comment (xiang90) [inline proxy/grpcproxy/cluster.go]: this import needs to be fixed. Put it together with the std lib import block.

**Comment by xiang90** [comment]
Lgtm after fixing the nits

File: clientv3/cluster.go (hunk 1 above)
Comment (jingyih) [reply clientv3/cluster.go]: Done.

**Comment by xiang90** [comment]
Let us merge the commits in pr by pr.

**Comment by jingyih** [comment]
> Let us merge the commits in pr by pr.

Can you clarify?

**Comment by xiang90** [comment]
@jingyih Let us split  #10645 to multiple PRs and merge them one by one. This is the first one that we will merge soon.

**Comment by jingyih** [comment]
> @jingyih Let us split #10645 to multiple PRs and merge them one by one. This is the first one that we will merge soon.

Got it.

File: proxy/grpcproxy/cluster.go (hunk 3 above)
Comment (jingyih) [reply proxy/grpcproxy/cluster.go]: Fixed.

**Comment by gyuho** [comment]
@jingyih Can you highlight those changes from this PR in our changelog? Thanks for the great work!

**Comment by jingyih** [comment]
> @jingyih Can you highlight those changes from this PR in our changelog? Thanks for the great work!

Sure I will update the changelog.

**Pull request merged**